libs/gl-client/src/signer/mod.rs

```diff
@@ -225,15 +230,55 @@ impl Signer {
                 if r.timestamp != 0 {
                     data.put_u64(r.timestamp);
                 }
-                data.put(&r.rune[..]);
+                let rune: Vec<u8> = r.rune.clone();
+                data.put(&rune[..]);
 
                 pk.verify(&data, &r.signature)
+                    .map_err(|e| anyhow!("signature verification failed: {}", e))?;
+
+                self.verify_rune(r.clone())
                     .map(|_| r)
-                    .map_err(|e| anyhow!("signature verification failed: {}", e))
+                    .map_err(|e| anyhow!("rune verification failed: {}", e))
             })
             .collect()
     }
 
+    /// Verifies that the public key of the request and the signers rune version
+    /// match the corresponding restrictions of the rune.
+    fn verify_rune(&self, request: crate::pb::PendingRequest) -> Result<(), anyhow::Error> {
+        let rune64 = general_purpose::URL_SAFE.encode(request.rune);
+        let rune = Rune::from_base64(&rune64)?;
+
+        // A valid gl-rune must contain a pubkey field as this  is bound to the
+        // signer. Against the rules of runes we do not accept a rune that has
+        // no restriction on a public key.
+        if !rune.to_string().contains("pubkey=") {
+            return Err(anyhow!("rune is missing pubkey field"));
+        }
```

Comment:
Is this not redundant with the check directly below? Doing a string search on the unparsed restrictions sounds like it can easily be tricked by the client just appending something that contains the substring you are searching for.

Reply:
To be fair, this check is pretty useless if an attacker gets access to an *unrestricted* rune. This check should just ensure that we do not use unrestricted runes by accident (in development, while testing etc):
The thing with runes being a set of restrictions is: If the restriction is not present, we would not check on it, so a rune that does not have a `pubkey` field present would pass the second check as we would simply skip the check. So the first check ensures that the second check will always be performed, while the second check ensures that the public key of the request matches the restriction of the rune, performing the check on the rune itself and not the unparsed restriction.

>Doing a string search on the unparsed restrictions sounds like it can easily be tricked by the client just appending something that contains the substring you are searching for.

That is correct and the reason why the length-extension-attack-feature of runes only allows them to be a set of restrictions, rather than allowances. But as said above, this is only a problem if an attacker has access to an unrestricted rune and this is the reason why we must be especially careful that we do not distribute runes that are not restricted to the `pubkey` of the device.